File: mgindb/mgindb.js
import WebSocket from 'ws'; // Make sure to install 'ws' package via npm

class MginDBClient {
    constructor(protocol = 'ws', host = '127.0.0.1', port = 6446, username = '', password = '') {
        this.uri = `${protocol}://${host}:${port}`; // WebSocket URI
        this.username = username; // Username for authentication
        this.password = password; // Password for authentication
        this.websocket = null; // WebSocket client
    }

    async connect() {
        /**
         * Connects to the MginDB server and authenticates the user.
         *
         * @throws Error If authentication fails.
         */
        this.websocket = new WebSocket(this.uri);

        // Wait for connection to open
        await new Promise((resolve, reject) => {
            this.websocket.on('open', resolve);
            this.websocket.on('error', reject);
        });

        const authData = JSON.stringify({ username: this.username, password: this.password });
        this.websocket.send(authData);

        // Await the first message which should be a welcome message
        const response = await new Promise(resolve => this.websocket.on('message', resolve));
        if (response !== "MginDB server connected... Welcome!") {
            throw new Error("Failed to authenticate: " + response);
        }
    }

    async sendCommand(command) {
        /**
         * Sends a command to the MginDB server and returns the response.
         *
         * @param {string} command The command to send.
         * @returns {Promise<string>} The server response.
         */
        if (!this.websocket || this.websocket.readyState !== WebSocket.OPEN) {
            await this.connect();
        }
        this.websocket.send(command);
        return new Promise(resolve => this.websocket.on('message', resolve));
    }

    async close() {
        /**
         * Closes the WebSocket connection to the MginDB server.
         */
        if (this.websocket) {
            this.websocket.close();
        }
    }

    set(key, value) {
        /**
         * Sets a value for a given key in the MginDB.
         *
         * @param {string} key The key to set.
         * @param {string} value The value to set.
         * @returns {Promise<string>} The server response.
         */
        return this.sendCommand(`SET ${key} ${value}`);
    }

    indices(action, key = null, value = null) {
        /**
         * Manages indices in the MginDB.
         *
         * @param {string} action The action to perform (e.g., LIST, ADD).
         * @param {string|null} key The key for the action.
         * @param {string|null} value The value for the action.
         * @returns {Promise<string>} The server response.
         */
        const command = `INDICES ${action}` + (key ? ` ${key}` : '') + (value ? ` ${value}` : '');
        return this.sendCommand(command);
    }

    incr(key, value) {
        /**
         * Increments a key's value by a given amount in the MginDB.
         *
         * @param {string} key The key to increment.
         * @param {string} value The amount to increment by.
         * @returns {Promise<string>} The server response.
         */
        return this.sendCommand(`INCR ${key} ${value}`);
    }

    decr(key, value) {
        /**
         * Decrements a key's value by a given amount in the MginDB.
         *
         * @param {string} key The key to decrement.
         * @param {string} value The amount to decrement by.
         * @returns {Promise<string>} The server response.
         */
        return this.sendCommand(`DECR ${key} ${value}`);
    }

    delete(key) {
        /**
         * Deletes a key from the MginDB.
         *
         * @param {string} key The key to delete.
         * @returns {Promise<string>} The server response.
         */
        return this.sendCommand(`DEL ${key}`);
    }

    query(key, queryString = null, options = null) {
        /**
         * Queries the MginDB with a given key and query string.
         *
         * @param {string} key The key to query.
         * @param {string|null} queryString The query string.
         * @param {string|null} options Additional options for the query.
         * @returns {Promise<string>} The server response.
         */
        const command = `QUERY ${key}` + (queryString ? ` ${queryString}` : '') + (options ? ` ${options}` : '');
        return this.sendCommand(command);
    }

    count(key) {
        /**
         * Counts the number of records for a given key in the MginDB.
         *
         * @param {string} key The key to count.
         * @returns {Promise<string>} The server response.
         */
        return this.sendCommand(`COUNT ${key}`);
    }

    schedule(action, cronOrKey = null, command = null) {
        /**
         * Schedules a task in the MginDB.
         *
         * @param {string} action The action to perform (e.g., ADD, REMOVE).
         * @param {string|null} cronOrKey The cron expression or key for the action.
         * @param {string|null} command The command to schedule.
         * @returns {Promise<string>} The server response.
         */
        const cmd = `SCHEDULE ${action}` + (cronOrKey ? ` ${cronOrKey}` : '') + (command ? ` ${command}` : '');
        return this.sendCommand(cmd);
    }

    sub(key) {
        /**
         * Subscribes to a key in the MginDB.
         *
         * @param {string} key The key to subscribe to.
         * @returns {Promise<string>} The server response.
         */
        return this.sendCommand(`SUB ${key}`);
    }

    unsub(key) {
        /**
         * Unsubscribes from a key in the MginDB.
         *
         * @param {string} key The key to unsubscribe from.
         * @returns {Promise<string>} The server response.
         */
        return this.sendCommand(`UNSUB ${key}`);
    }
}

export default MginDBClient;

File: mgindb/mgindb.go
package main

import (
    "encoding/json"
    "fmt"
    "log"
    "net/url"
    "sync"
    "time"

    "github.com/gorilla/websocket"
)

type MginDBClient struct {
    uri        string
    username   string
    password   string
    connection *websocket.Conn
    mutex      sync.Mutex
}

type AuthData struct {
    Username string `json:"username"`
    Password string `json:"password"`
}

func NewMginDBClient(protocol, host string, port int, username, password string) *MginDBClient {
    uri := fmt.Sprintf("%s://%s:%d", protocol, host, port)
    return &MginDBClient{uri: uri, username: username, password: password}
}

func (client *MginDBClient) Connect() error {
    u, err := url.Parse(client.uri)
    if err != nil {
        return err
    }

    client.mutex.Lock()
    defer client.mutex.Unlock()

    c, _, err := websocket.DefaultDialer.Dial(u.String(), nil)
    if err != nil {
        return err
    }
    client.connection = c

    authData := AuthData{Username: client.username, Password: client.password}
    authDataJson, err := ujson.Marshal(authData)
    if err != nil {
        return err
    }

    err = client.connection.WriteMessage(websocket.TextMessage, authDataJson)
    if err != nil {
        return err
    }

    _, message, err := client.connection.ReadMessage()
    if err != nil {
        return err
    }

    if string(message) != "MginDB server connected... Welcome!" {
        return fmt.Errorf("failed to authenticate: %s", message)
    }

    return nil
}

func (client *MginDBClient) sendCommand(command string) (string, error) {
    client.mutex.Lock()
    defer client.mutex.Unlock()

    if client.connection == nil {
        if err := client.Connect(); err != nil {
            return "", err
        }
    }

    err := client.connection.WriteMessage(websocket.TextMessage, []byte(command))
    if err != nil {
        return "", err
    }

    _, message, err := client.connection.ReadMessage()
    if err != nil {
        return "", err
    }

    return string(message), nil
}

func (client *MginDBClient) Set(key, value string) (string, error) {
    return client.sendCommand(fmt.Sprintf("SET %s %s", key, value))
}

func (client *MginDBClient) Indices(action, key, value string) (string, error) {
    return client.sendCommand(fmt.Sprintf("INDICES %s %s %s", action, key, value))
}

func (client *MginDBClient) Incr(key, value string) (string, error) {
    return client.sendCommand(fmt.Sprintf("INCR %s %s", key, value))
}

func (client *MginDBClient) Decr(key, value string) (string, error) {
    return client.sendCommand(fmt.Sprintf("DECR %s %s", key, value))
}

func (client *MginDBClient) Delete(key string) (string, error) {
    return client.sendCommand(fmt.Sprintf("DEL %s", key))
}

func (client *MginDBClient) Query(key, queryString, options string) (string, error) {
    return client.sendCommand(fmt.Sprintf("QUERY %s %s %s", key, queryString, options))
}

func (client *MginDBClient) Count(key string) (string, error) {
    return client.sendCommand(fmt.Sprintf("COUNT %s", key))
}

func (client *MginDBClient) Schedule(action, cronOrKey, command string) (string, error) {
    return client.sendCommand(fmt.Sprintf("SCHEDULE %s %s %s", action, cronOrKey, command))
}

func (client *MginDBClient) Sub(key string) (string, error) {
    return client.sendCommand(fmt.Sprintf("SUB %s", key))
}

func (client *MginDBClient) Unsub(key string) (string, error) {
    return client.sendCommand(fmt.Sprintf("UNSUB %s", key))
}

func (client *MginDBClient) Close() error {
    client.mutex.Lock()
    defer client.mutex.Unlock()

    if client.connection != nil {
        err := client.connection.Close()
        client.connection = nil
        return err
    }
    return nil
}
